Framework/PythonInterface/plugins/algorithms/PelicanCrystalProcessing.py

```diff
@@ -0,0 +1,219 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+import os
+import h5py
+
+from mantid import config
+from mantid.kernel import (Direction, StringArrayMandatoryValidator, StringArrayProperty, CompositeValidator)
+from mantid.api import (PythonAlgorithm, FileProperty, FileAction, Progress)
+from mantid.simpleapi import *  # noqa
+
+
+def seq_to_list(iseqn):
+    # convert a comma separated range of numbers returned as a list
+    # first clean all whitespaces
+    seqn = iseqn.replace(' ', '')
+    nlist = []
+    sqlist = seqn.split(',')
+    for rg in sqlist:
+        if rg == '':
+            continue
+        ss = rg.split('-')
+        try:
+            lo = int(ss[0])
+            hi = int(ss[-1])
+            if len(ss) == 1:
+                nlist.append(lo)
+            else:
+                for i in range(lo, hi+1):
+                    nlist.append(i)
+        except ValueError:
+            raise RuntimeError('Unexpected run sequence: {}'.format(seqn))
+
+    return nlist
+
+
+def cycle_and_runs(run_seq):
+    ss = run_seq.split('::')
+    if len(ss) == 2:
+        return int(ss[0]), seq_to_list(ss[1])
+    else:
+        return None, seq_to_list(ss[0])
+
+
+def expand_as_cycle_runs(sample_runs):
+    cycle_runs = []
+    for run_seq in sample_runs.split(';'):
+        cycle, runs = cycle_and_runs(run_seq)
+        for run in runs:
+            cycle_runs.append((cycle, run))
+    return cycle_runs
+
+
+def copy_datset_nodes(src_hdf, dst_hdf, tags):
+    with h5py.File(dst_hdf, 'r+') as f_dst:
+        with h5py.File(src_hdf, 'r') as f_src:
+            for tag in tags:
+                sval = f_src[tag][()]
+                dset = f_dst[tag]
+                dset[...] = sval
+
+
+class PelicanCrystalProcessing(PythonAlgorithm):
```

Comment:
Can we inherit from `DataProcessorAlgorithm` instead? (https://github.com/mantidproject/mantid/search?l=Python&q=DataProcessorAlgorithm)

This will mean that the history of the internal algorithm calls can be "unwound" if someone wants to see the workflow in more detail.

Reply:
Done